django_project/frontend/api_views/map.py

```diff
@@ -35,11 +46,93 @@
 from frontend.serializers.property import (
     PropertySerializer
 )
-from frontend.utils.organisation import (
-    get_current_organisation_id
-)
+from stakeholder.models import OrganisationUser
+from frontend.utils.organisation import get_current_organisation_id
+
 
 User = get_user_model()
+PROVINCE_LAYER_ZOOMS = (5, 8)
+PROPERTIES_LAYER_ZOOMS = (10, 24)
+PROPERTIES_POINT_LAYER_ZOOMS = (5, 24)
+
+
+def should_generate_layer(z: int, zoom_configs: Tuple[int, int]) -> bool:
+    """Return True if layer should be generated."""
+    return z >= zoom_configs[0] and z <= zoom_configs[1]
+
+
+class MapSessionBase(APIView):
+    """Base class for map filter session."""
+
+    def can_view_properties_layer(self):
+        return self.request.user.has_perm(
+            'sawps.can_view_map_properties_layer')
+
+    def can_view_province_layer(self):
+        return self.request.user.has_perm('sawps.can_view_map_province_layer')
+
+    def generate_session(self):
+        """
+        Generate map filter session from POST data.
+
+        Session will have expiry in 6 hours after creation and
+        this expiry date is updated when the session filter is changed.
+        """
+        session_uuid = self.request.GET.get('session', None)
+        session = MapSession.objects.filter(uuid=session_uuid).first()
+        filter_species = self.get_species_filter()
+        if session is None:
+            session = MapSession.objects.create(
+                user=self.request.user,
+                created_date=timezone.now(),
+                expired_date=timezone.now() + datetime.timedelta(hours=6)
+            )
+        session.species = filter_species
+        session.expired_date = timezone.now() + datetime.timedelta(hours=6)
+        session.save(update_fields=['species', 'expired_date'])
+        if self.can_view_properties_layer():
+            generate_map_view(
+                session, False,
```

Comment:
Perhaps we could add a variable to check if it's a province layer, so we don't need to write the generate_map_view function twice.

Reply:
@dimasciput, actually the system can generate two materialized views, 1 for properties and 1 for province. This can happen for super user or user that belongs to both decision maker and data scientiest.